# Data Ingestion

**Objective:** Load the IBM HR Analytics dataset and perform initial exploration

**Dataset:** IBM HR Analytics Employee Attrition & Performance
- Source: Kaggle
- Records: 1,470 employees
- Features: 35 columns

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
# Load the dataset
print("Loading IBM HR Analytics dataset...")

df = pd.read_csv('../data/raw/hr_data.csv')

print(f"Data loaded successfully!")
print(f"Total Employees: {len(df)}")
print(f"Total Features: {df.shape[1]}")

Loading IBM HR Analytics dataset...
Data loaded successfully!
Total Employees: 1470
Total Features: 35


In [3]:
# Display first few rows
print("\nFirst 5 rows of the dataset:")
print("="*80)
df.head()


First 5 rows of the dataset:


Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [4]:
# Check all column names
print("All Columns in Dataset:")
print("="*80)
for i, col in enumerate(df.columns, 1):
    print(f"{i:2d}. {col}")

All Columns in Dataset:
 1. Age
 2. Attrition
 3. BusinessTravel
 4. DailyRate
 5. Department
 6. DistanceFromHome
 7. Education
 8. EducationField
 9. EmployeeCount
10. EmployeeNumber
11. EnvironmentSatisfaction
12. Gender
13. HourlyRate
14. JobInvolvement
15. JobLevel
16. JobRole
17. JobSatisfaction
18. MaritalStatus
19. MonthlyIncome
20. MonthlyRate
21. NumCompaniesWorked
22. Over18
23. OverTime
24. PercentSalaryHike
25. PerformanceRating
26. RelationshipSatisfaction
27. StandardHours
28. StockOptionLevel
29. TotalWorkingYears
30. TrainingTimesLastYear
31. WorkLifeBalance
32. YearsAtCompany
33. YearsInCurrentRole
34. YearsSinceLastPromotion
35. YearsWithCurrManager


In [5]:
# Dataset info
print("\nDataset Information:")
print("="*80)
df.info()


Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  Jo

In [6]:
# Check for missing values
print("\nMissing Values Check:")
print("="*80)

missing = df.isnull().sum()
if missing.sum() == 0:
    print("Great! No missing values found in the dataset.")
else:
    print("Missing values found:")
    print(missing[missing > 0])


Missing Values Check:
Great! No missing values found in the dataset.


In [7]:
# Check for duplicates
print("\nDuplicate Records Check:")
print("="*80)

duplicates = df.duplicated().sum()
if duplicates == 0:
    print("No duplicate records found.")
else:
    print(f"Found {duplicates} duplicate records.")


Duplicate Records Check:
No duplicate records found.


In [8]:
# Basic statistics for numeric columns
print("\nStatistical Summary (Numeric Features):")
print("="*80)
df.describe().round(2)


Statistical Summary (Numeric Features):


Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
count  1470.00    1470.00           1470.00    1470.00         1470.0   
mean     36.92     802.49              9.19       2.91            1.0   
std       9.14     403.51              8.11       1.02            0.0   
min      18.00     102.00              1.00       1.00            1.0   
25%      30.00     465.00              2.00       2.00            1.0   
50%      36.00     802.00              7.00       3.00            1.0   
75%      43.00    1157.00             14.00       4.00            1.0   
max      60.00    1499.00             29.00       5.00            1.0   

       EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
count         1470.00                  1470.00     1470.00         1470.00   
mean          1024.87                     2.72       65.89            2.73   
std            602.02                     1.09       20.33            0.71   
min              1.00                     1.00       30.00            1.00   
25%            491.25                     2.00       48.00            2.00   
50%           1020.50                     3.00       66.00            3.00   
75%           1555.75                     4.00       83.75            3.00   
max           2068.00                     4.00      100.00            4.00   

       JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count   1470.00  ...                   1470.00         1470.0   
mean       2.06  ...                      2.71           80.0   
std        1.11  ...                      1.08            0.0   
min        1.00  ...                      1.00           80.0   
25%        1.00  ...                      2.00           80.0   
50%        2.00  ...                      3.00           80.0   
75%        3.00  ...                      4.00           80.0   
max        5.00  ...                      4.00           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count           1470.00            1470.00                1470.00   
mean               0.79              11.28                   2.80   
std                0.85               7.78                   1.29   
min                0.00               0.00                   0.00   
25%                0.00               6.00                   2.00   
50%                1.00              10.00                   3.00   
75%                1.00              15.00                   3.00   
max                3.00              40.00                   6.00   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count          1470.00         1470.00             1470.00   
mean              2.76            7.01                4.23   
std               0.71            6.13                3.62   
min               1.00            0.00                0.00   
25%               2.00            3.00                2.00   
50%               3.00            5.00                3.00   
75%               3.00            9.00                7.00   
max               4.00           40.00               18.00   

       YearsSinceLastPromotion  YearsWithCurrManager  
count                  1470.00               1470.00  
mean                      2.19                  4.12  
std                       3.22                  3.57  
min                       0.00                  0.00  
25%                       0.00                  2.00  
50%                       1.00                  3.00  
75%                       3.00                  7.00  
max                      15.00                 17.00  

[8 rows x 26 columns]

In [9]:
# Check target variable distribution
print("\nTarget Variable Distribution (Attrition):")
print("="*80)

attrition_counts = df['Attrition'].value_counts()
attrition_pct = df['Attrition'].value_counts(normalize=True) * 100

print(attrition_counts)
print(f"\nAttrition Rate: {attrition_pct['Yes']:.2f}%")
print(f"Retention Rate: {attrition_pct['No']:.2f}%")


Target Variable Distribution (Attrition):
Attrition
No     1233
Yes     237
Name: count, dtype: int64

Attrition Rate: 16.12%
Retention Rate: 83.88%


In [10]:
# Check categorical columns
print("\nCategorical Features Summary:")
print("="*80)

categorical_cols = df.select_dtypes(include=['object']).columns

for col in categorical_cols:
    print(f"\n{col}:")
    print(df[col].value_counts())
    print(f"Unique values: {df[col].nunique()}")


Categorical Features Summary:

Attrition:
Attrition
No     1233
Yes     237
Name: count, dtype: int64
Unique values: 2

BusinessTravel:
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64
Unique values: 3

Department:
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64
Unique values: 3

EducationField:
EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64
Unique values: 6

Gender:
Gender
Male      882
Female    588
Name: count, dtype: int64
Unique values: 2

JobRole:
JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Direct

In [11]:
# Save a copy for next steps
print("\nSaving data for next notebook...")

# Save to external folder (we'll enrich this later)
df.to_csv('../data/external/enriched_data.csv', index=False)

print("Data saved to: data/external/enriched_data.csv")
print("\nNotebook 01 Complete! Ready for data cleaning.")


Saving data for next notebook...
Data saved to: data/external/enriched_data.csv

Notebook 01 Complete! Ready for data cleaning.


---

## Findings from Data Ingestion:

1. **Dataset Size:** 1,470 employees with 35 features
2. **Data Quality:** No missing values, no duplicates
3. **Attrition Rate:** 16.12% (industry standard is 10-15%)
4. **Data Types:** Mix of numeric and categorical
5. **Ready for Cleaning:** Minimal preprocessing needed

---

**Next Step:** Proceed to `02_data_cleaning.ipynb`